packages/headless/src/controllers/knowledge/generated-answer/headless-answerapi-generated-answer.ts

```diff
@@ -211,3 +242,128 @@ function loadAnswerApiReducers(
   engine.addReducers({[answerApi.reducerPath]: answerApi.reducer, query});
   return true;
 }
+
+const buildAdvancedSearchQueryParams = (state: StateNeededByAnswerAPI) => {
+  const advancedSearchQueryParams = selectAdvancedSearchQueries(state);
+  const mergedCq = mergeAdvancedCQParams(state);
+
+  return {
+    ...advancedSearchQueryParams,
+    ...(mergedCq && {cq: mergedCq}),
+  };
+};
+
+const mergeAdvancedCQParams = (state: StateNeededByAnswerAPI) => {
+  const activeTabExpression = selectActiveTabExpression(state.tabSet);
+  const filterExpressions = selectStaticFilterExpressions(state);
+  const {cq} = selectAdvancedSearchQueries(state);
+
+  return [activeTabExpression, ...filterExpressions, cq]
+    .filter((expression) => !!expression)
+    .join(' AND ');
+};
+
+const getNumberOfResultsWithinIndexLimit = (state: StateNeededByAnswerAPI) => {
+  if (!state.pagination) {
+    return undefined;
+  }
+
+  const isOverIndexLimit =
+    state.pagination.firstResult + state.pagination.numberOfResults >
+    maximumNumberOfResultsFromIndex;
+
+  if (isOverIndexLimit) {
+    return maximumNumberOfResultsFromIndex - state.pagination.firstResult;
+  }
+  return state.pagination.numberOfResults;
+};
+
+export const constructAnswerQueryParams = (
```

Comment:
This one could be defined under `features/generated-answer/generated-answer-request.ts`

Reply:
sure I can move it

Reply:
I thought I mentioned that in another comment, sorry if I did and I'm repeating my self here.
But the goal is to avoid adding a lot of complexity/logic to the controller file. this file should remain simple with one precise goal, offering a controller that gives access to the state and a set of methods.

All the logic to constructAnswerQueryParams will be better located under `features/generated-answer/generated-answer-request.ts`